src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,177 @@
+using System.Threading.Tasks;
+
+using Analyzer.Utilities;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
+{
+    // TODO: VisualBasic as well?
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class StringBuilderAppendShouldNotTakeSubstringFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray
+            .Create(
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId,
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdTwoParameterId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            SyntaxNode root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            SyntaxNode node = root.FindNode(context.Span);
+
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                if (diagnostic.Id == StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId)
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithOneParameterFix,
+                                async ctx => await FixCodeOneParameter(context.Document, root, node, ctx).ConfigureAwait(false),
+                                equivalenceKey: MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithOneParameterFix),
+                            diagnostic);
+                }
+                else
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterFix,
+                                async ctx => await FixCodeTwoParameters(context.Document, root, node, ctx).ConfigureAwait(false),
+                                equivalenceKey: MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterFix),
+                            diagnostic);
+                }
+            }
+        }
+
+        private static async Task<Document> FixCodeOneParameter(
+            Document document,
+            SyntaxNode root,
+            SyntaxNode nodeToFix,
+            CancellationToken cancellationToken)
+        {
+            var typedNodeToFix = (InvocationExpressionSyntax)nodeToFix;
```

Comment:
my comments seem to have been lost.  So, instead of casting to a C# specific type.  yo ucan use IOperatoin here and get the actual IOperation IInvocationOperation type and work with that.  it will work for VB and C# then.

Reply:
okay, so creating a semanticModel by using document.GetSemanticModel() first, gotcha... thanks.